File: Winter2022/examples/java/Lecture 6/ThreadingHandlerRunnable/app/src/main/java/mobiledev/unb/ca/threadinghandlerrunnable/HandlerRunnableActivity.java
package mobiledev.unb.ca.threadinghandlerrunnable;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

public class HandlerRunnableActivity extends Activity {
    private final static String PROGRESS_BAR_PROGRESS_KEY = "PROGRESS_BAR_PROGRESS_KEY";
    private final static String PROGRESS_BAR_VISIBLE_KEY = "PROGRESS_BAR_VISIBLE_KEY";
    private final static String BITMAP_KEY = "BITMAP_KEY";

    private LoadIconTask mLoadIconTask;
    private ImageView mImageView;
    private ProgressBar mProgressBar;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        mImageView = findViewById(R.id.imageView);
        mProgressBar = findViewById(R.id.progressBar);

        if (null != savedInstanceState) {
            mProgressBar.setVisibility(savedInstanceState.getInt(PROGRESS_BAR_VISIBLE_KEY));
            mProgressBar.setProgress(savedInstanceState.getInt(PROGRESS_BAR_PROGRESS_KEY));
            mImageView.setImageBitmap((Bitmap) savedInstanceState.getParcelable(BITMAP_KEY));
            mLoadIconTask = ((LoadIconTask) getLastNonConfigurationInstance());
            if (null != mLoadIconTask) {
                mLoadIconTask.setProgressBar(mProgressBar)
                        .setImageView(mImageView);
            }
        }
    }

    public void onClickLoadButton(View v) {
        v.setEnabled(false);
        mLoadIconTask = new LoadIconTask(getApplicationContext())
                .setImageView(mImageView)
                .setProgressBar(mProgressBar);
        mLoadIconTask.start();
    }


    public void onClickOtherButton(View v) {
        Toast.makeText(HandlerRunnableActivity.this, "I'm Working",
                Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(PROGRESS_BAR_VISIBLE_KEY, mProgressBar.getVisibility());
        outState.putInt(PROGRESS_BAR_PROGRESS_KEY, mProgressBar.getProgress());

        if (null != mImageView.getDrawable()) {
            Bitmap bm = ((BitmapDrawable) mImageView.getDrawable()).getBitmap();
            outState.putParcelable(BITMAP_KEY, bm);
        }
    }

    @Override
    public Object onRetainNonConfigurationInstance() {
        return mLoadIconTask;
    }
}

File: Fall2023/examples/Lecture 6/ThreadingHandlerRunnable/app/src/main/java/mobiledev/unb/ca/threading/handlerrunnable/HandlerRunnableActivity.kt
package mobiledev.unb.ca.threading.handlerrunnable

import android.app.Activity
import android.widget.ProgressBar
import android.os.Bundle
import android.os.Parcelable
import android.graphics.Bitmap
import android.widget.Toast
import android.graphics.drawable.BitmapDrawable
import android.view.View
import android.widget.ImageView

class HandlerRunnableActivity : Activity() {
    private var mLoadIconTask: LoadIconTask? = null
    private var mImageView: ImageView? = null
    private var mProgressBar: ProgressBar? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        mImageView = findViewById(R.id.imageView)
        mProgressBar = findViewById(R.id.progressBar)

        if (null != savedInstanceState) {
            mProgressBar!!.visibility = savedInstanceState.getInt(PROGRESS_BAR_VISIBLE_KEY)
            mProgressBar!!.progress = savedInstanceState.getInt(PROGRESS_BAR_PROGRESS_KEY)
            mImageView!!.setImageBitmap(savedInstanceState.getParcelable<Parcelable>(BITMAP_KEY) as Bitmap?)
            mLoadIconTask = lastNonConfigurationInstance as LoadIconTask?
            if (null != mLoadIconTask) {
                mLoadIconTask!!.setProgressBar(mProgressBar)
                    .setImageView(mImageView)
            }
        }
    }

    fun onClickLoadButton(v: View) {
        v.isEnabled = false
        mLoadIconTask = LoadIconTask(applicationContext)
            .setImageView(mImageView)
            .setProgressBar(mProgressBar)
        mLoadIconTask!!.start()
    }

    fun onClickOtherButton(v: View?) {
        Toast.makeText(this@HandlerRunnableActivity, "I'm Working",
            Toast.LENGTH_SHORT).show()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(PROGRESS_BAR_VISIBLE_KEY, mProgressBar!!.visibility)
        outState.putInt(PROGRESS_BAR_PROGRESS_KEY, mProgressBar!!.progress)
        if (null != mImageView!!.drawable) {
            val bm = (mImageView!!.drawable as BitmapDrawable).bitmap
            outState.putParcelable(BITMAP_KEY, bm)
        }
    }

    override fun onRetainNonConfigurationInstance(): Any {
        return mLoadIconTask!!
    }

    companion object {
        private const val PROGRESS_BAR_PROGRESS_KEY = "PROGRESS_BAR_PROGRESS_KEY"
        private const val PROGRESS_BAR_VISIBLE_KEY = "PROGRESS_BAR_VISIBLE_KEY"
        private const val BITMAP_KEY = "BITMAP_KEY"
    }
}